File: Algorithms/src/3568. Minimum Moves to Clean the Classroom/Code/solution.py
from typing import List
from collections import deque

class Solution:
    def minMoves(self, classroom: List[str], energy: int) -> int:
        m = len(classroom)
        n = len(classroom[0])

        # id[r][c] stores the bit assigned to a litter cell.
        id = [[-1] * n for _ in range(m)]

        k = 0
        sr = 0
        sc = 0

        # Find the starting position and assign bits to all litter cells.
        for r in range(m):
            for c in range(n):
                if classroom[r][c] == 'S':
                    sr = r
                    sc = c
                elif classroom[r][c] == 'L':
                    id[r][c] = k
                    k += 1

        # If there is no litter, the task is already complete.
        if k == 0:
            return 0

        # This mask has all k litter bits turned on.
        total_mask = (1 << k) - 1

        # best[r][c][mask] stores the maximum energy seen
        # for this position and collected-litter mask.
        best = [
            [
                [-1] * (1 << k)
                for _ in range(n)
            ]
            for _ in range(m)
        ]

        # BFS state is (row, column, mask, remaining energy, moves).
        queue = deque()

        # Start at S with no collected litter and full energy.
        best[sr][sc][0] = energy
        queue.append((sr, sc, 0, energy, 0))

        # Four possible movement directions.
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        while queue:
            # Get the next state in BFS order.
            r, c, mask, e, moves = queue.popleft()

            # Try moving to all four neighboring cells.
            for dr, dc in directions:
                nr = r + dr
                nc = c + dc

                # Ignore positions outside the classroom.
                if nr < 0 or nr >= m or nc < 0 or nc >= n:
                    continue

                # Obstacles cannot be entered.
                if classroom[nr][nc] == 'X':
                    continue

                # Every movement uses one unit of energy.
                ne = e - 1

                # We cannot move if energy becomes negative.
                if ne < 0:
                    continue

                nmask = mask

                # Reset energy when entering an R cell.
                if classroom[nr][nc] == 'R':
                    ne = energy

                # Mark the litter as collected by setting its bit.
                if classroom[nr][nc] == 'L':
                    nmask |= 1 << id[nr][nc]

                # All litter has been collected, so return the move count.
                if nmask == total_mask:
                    return moves + 1

                # If an equal or stronger state already exists,
                # this new state cannot give us a better path.
                if ne <= best[nr][nc][nmask]:
                    continue

                # Keep the maximum energy for this position and mask.
                best[nr][nc][nmask] = ne

                # Add the improved state to the BFS queue.
                queue.append((nr, nc, nmask, ne, moves + 1))

        # BFS could not find a path that collects all litter.
        return -1

File: Algorithms/src/3568. Minimum Moves to Clean the Classroom/Code/solution.cpp
class Solution
{
public:
    int minMoves(vector<string> &classroom, int energy)
    {
        int m = classroom.size();
        int n = classroom[0].size();

        // Give every litter cell a unique bit position.
        vector<vector<int>> id(m, vector<int>(n, -1));

        int k = 0;
        int sr = 0, sc = 0;

        // Find the starting position and assign IDs to all litter cells.
        for (int r = 0; r < m; r++)
        {
            for (int c = 0; c < n; c++)
            {
                if (classroom[r][c] == 'S')
                {
                    sr = r;
                    sc = c;
                }
                else if (classroom[r][c] == 'L')
                {
                    id[r][c] = k++;
                }
            }
        }

        // If there is no litter, the student is already done.
        if (k == 0)
            return 0;

        int totalMask = (1 << k) - 1;

        // bestEnergy[r][c][mask] stores the maximum energy seen
        // at this position after collecting exactly this mask.
        vector<vector<vector<int>>> best(
            m, vector<vector<int>>(n, vector<int>(1 << k, -1)));

        // Each BFS state contains position, collected litter mask,
        // remaining energy, and number of moves used.
        struct State
        {
            int r, c, mask, e, moves;
        };

        queue<State> q;

        // Initially, no litter is collected and full energy is available.
        best[sr][sc][0] = energy;
        q.push({sr, sc, 0, energy, 0});

        // Four possible movement directions.
        int dr[] = {-1, 1, 0, 0};
        int dc[] = {0, 0, -1, 1};

        while (!q.empty())
        {
            State cur = q.front();
            q.pop();

            // Try moving in all four directions.
            for (int d = 0; d < 4; d++)
            {
                int nr = cur.r + dr[d];
                int nc = cur.c + dc[d];

                // Ignore positions outside the classroom.
                if (nr < 0 || nr >= m || nc < 0 || nc >= n)
                    continue;

                // The student cannot move through obstacles.
                if (classroom[nr][nc] == 'X')
                    continue;

                // Every movement consumes one unit of energy.
                int ne = cur.e - 1;

                // The student cannot make a move without energy.
                if (ne < 0)
                    continue;

                int nmask = cur.mask;

                // Reset the energy immediately after entering an R cell.
                if (classroom[nr][nc] == 'R')
                {
                    ne = energy;
                }

                // If this cell contains litter, mark its bit as collected.
                if (classroom[nr][nc] == 'L')
                {
                    nmask |= (1 << id[nr][nc]);
                }

                // All litter has been collected, so this is the answer.
                if (nmask == totalMask)
                {
                    return cur.moves + 1;
                }

                // If we already reached this position with the same
                // mask and at least this much energy, this state is useless.
                if (ne <= best[nr][nc][nmask])
                    continue;

                // Keep only the strongest energy value for this state.
                best[nr][nc][nmask] = ne;

                // Add the improved state to BFS.
                q.push({nr, nc, nmask, ne, cur.moves + 1});
            }
        }

        // BFS finished without collecting all litter.
        return -1;
    }
};